File: core/src/com/oni/donjon/map/Tile.java
package com.oni.donjon.map;

import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.physics.box2d.*;
import com.badlogic.gdx.utils.NumberUtils;
import com.oni.donjon.screen.GameScreen;

import java.util.Objects;

/**
 * @author Daniel Chesters (on 22/05/14).
 */
public class Tile {
    public static final float SIZE = 32f;
    private TileType type;
    private Rectangle rectangle;
    private boolean know;
    private Body body;

    public boolean isKnow() {
        return know;
    }

    public void setKnow(boolean know) {
        this.know = know;
    }

    public TileType getType() {
        return type;
    }

    public void setType(TileType type) {
        this.type = type;
    }

    public Rectangle getRectangle() {
        return rectangle;
    }

    @Override public String toString() {
        return "Tile{" +
            "type=" + type +
            ", rectangle=" + rectangle +
            ", know=" + know +
            '}';
    }

    public Tile() {
        this.rectangle = new Rectangle();
    }

    public Tile(float x, float y, TileType type, boolean know, World world) {
        this.rectangle = new Rectangle(x, y, SIZE, SIZE);
        this.type = type;
        this.know = know;
        if (type.getCategoryBits() != GameScreen.NOTHING_BIT) {
            this.body = createBody(world);
        }
    }

    public Body createBody(World world) {
        BodyDef bodyDef = new BodyDef();
        bodyDef.type = BodyDef.BodyType.StaticBody;
        bodyDef.position.set(rectangle.x * SIZE + rectangle.getWidth() / 2,
            rectangle.y * SIZE + rectangle.getHeight() / 2);
        Body body = world.createBody(bodyDef);

        PolygonShape polygonShape = new PolygonShape();
        polygonShape.setAsBox(rectangle.width / 2, rectangle.height / 2);
        FixtureDef fixtureDef = new FixtureDef();
        fixtureDef.shape = polygonShape;
        fixtureDef.filter.categoryBits = type.getCategoryBits();
        fixtureDef.filter.maskBits = GameScreen.PLAYER_BIT | GameScreen.LIGHT_BIT;
        body.createFixture(fixtureDef);
        polygonShape.dispose();

        return body;
    }

    public Body getBody() {
        return body;
    }

    public void setBody(Body body) {
        this.body = body;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        Tile tile = (Tile) o;
        return know == tile.know &&
            type == tile.type &&
            Objects.equals(rectangle, tile.rectangle) &&
            Objects.equals(body, tile.body);
    }

    @Override
    public int hashCode() {
        return Objects.hash(type, rectangle, know, body);
    }
}

File: core/src/main/kotlin/com/oni/donjon/map/Tile.kt
package com.oni.donjon.map

import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.physics.box2d.Body
import com.badlogic.gdx.physics.box2d.World
import com.oni.donjon.screen.GameScreen
import ktx.box2d.body
import ktx.box2d.box
import ktx.math.vec2
import kotlin.experimental.or

/**
 * @author Daniel Chesters (on 22/05/14).
 */
data class Tile(
    val x: Float = -1f,
    val y: Float = -1f,
    var type: TileType,
    var isKnow: Boolean = false,
    private val world: World
) {
    var rectangle: Rectangle
    lateinit var body: Body

    companion object {
        const val SIZE = 32f
    }

    init {
        if (x < 0 || y < 0) {
            this.rectangle = Rectangle()
        } else {
            this.rectangle = Rectangle(x, y, SIZE, SIZE)
            if (type.categoryBits != GameScreen.NOTHING_BIT) {
                this.body = createBody(world)
            }
        }
    }

    fun createBody(world: World): Body {
        return world.body {
            val realX = rectangle.x * SIZE + rectangle.getWidth() / 2
            val realY = rectangle.y * SIZE + rectangle.getHeight() / 2
            box(
                width = rectangle.width,
                height = rectangle.height,
                position = vec2(
                    realX,
                    realY
                )
            ) {
                filter.categoryBits = this@Tile.type.categoryBits
                filter.maskBits = GameScreen.PLAYER_BIT or GameScreen.LIGHT_BIT
            }
        }
    }
}
